sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLPlannerImpl.java

```diff
@@ -130,30 +135,57 @@ public RelRoot rel(SqlNode sqlNode) throws RelConversionException {
   }
 
   public RelRoot rel(String sql, QueryParameters params) {
-    this.cluster = RelOptCluster.create(planner, new RexBuilder(typeFactory));
-    this.expressionConverter = new ExpressionConverter(cluster, params);
+    RelOptCluster cluster = RelOptCluster.create(planner, new RexBuilder(typeFactory));
 
     QueryTrait trait = new QueryTrait();
 
-    // Set up table providers that need to be pre-registered
-    // TODO(https://issues.apache.org/jira/browse/BEAM-8817): share this logic between dialects
-    List<List<String>> tables = Analyzer.extractTableNamesFromStatement(sql);
-    TableResolution.registerTables(this.defaultSchemaPlus, tables);
-
-    ResolvedStatement statement =
+    SqlAnalyzer analyzer =
         SqlAnalyzer.getBuilder()
             .withQueryParams(params)
             .withQueryTrait(trait)
             .withCalciteContext(config.getContext())
             .withTopLevelSchema(defaultSchemaPlus)
             .withTypeFactory((JavaTypeFactory) cluster.getTypeFactory())
-            .analyze(sql);
+            .build();
+
+    AnalyzerOptions options = SqlAnalyzer.initAnalyzerOptions(params);
+
+    // Set up table providers that need to be pre-registered
+    List<List<String>> tables = analyzer.extractTableNames(sql, options);
+    TableResolution.registerTables(this.defaultSchemaPlus, tables);
+    SimpleCatalog catalog =
+        analyzer.createPopulatedCatalog(defaultSchemaPlus.getName(), options, tables);
+
+    ImmutableMap.Builder<String, ResolvedCreateFunctionStmt> udfBuilder = ImmutableMap.builder();
+
+    ResolvedStatement statement;
+    ParseResumeLocation parseResumeLocation = new ParseResumeLocation(sql);
+    do {
+      statement = analyzer.analyzeNextStatement(parseResumeLocation, options, catalog);
+      if (statement.nodeKind() == RESOLVED_CREATE_FUNCTION_STMT) {
+        ResolvedCreateFunctionStmt createFunctionStmt = (ResolvedCreateFunctionStmt) statement;
+        // ResolvedCreateFunctionStmt does not include the full function name, so build it here.
```

Comment:
Can you clarify what is "full function name"?

Reply:
ResolvedCreateFunctionStmt contains the path as a list of strings, while we need the whole path as a single string.